crates/core/tedge_api/src/plugin.rs

```diff
@@ -0,0 +1,619 @@
+    ///
+    /// # Note
+    ///
+    /// This function _must not_ block.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use tedge_api::plugin::BuiltPlugin;
+    /// # use tedge_api::PluginConfiguration;
+    /// # use tedge_api::Plugin;
+    /// # use tedge_api::PluginBuilder;
+    /// # use tedge_api::PluginDirectory;
+    ///
+    /// #[derive(Debug)]
+    /// struct MyMessage;
+    /// impl tedge_api::plugin::Message for MyMessage {
+    ///     type Reply = tedge_api::message::NoReply;
+    /// }
+    ///
+    ///
+    /// struct MyPluginBuilder;
+    /// struct MyPlugin; // + some impl Plugin for MyPlugin
+    /// # #[async_trait::async_trait]
+    /// # impl Plugin for MyPlugin {
+    /// #     async fn setup(&mut self) -> Result<(), tedge_api::error::PluginError> {
+    /// #         unimplemented!()
+    /// #     }
+    /// #     async fn shutdown(&mut self) -> Result<(), tedge_api::error::PluginError> {
+    /// #         unimplemented!()
+    /// #     }
+    /// # }
+    ///
+    /// #[async_trait::async_trait]
+    /// impl tedge_api::plugin::Handle<MyMessage> for MyPlugin {
+    ///     async fn handle_message(
+    ///         &self,
+    ///         _message: MyMessage,
+    ///         _sender: tedge_api::address::ReplySender<tedge_api::message::NoReply>,
+    ///     ) -> Result<(), tedge_api::error::PluginError> {
+    ///         // implementation...
+    /// #       unimplemented!()
+    ///     }
+    /// }
+    ///
+    /// #[async_trait::async_trait]
+    /// impl<PD: PluginDirectory> PluginBuilder<PD> for MyPluginBuilder {
+    ///     async fn instantiate(
+    ///         &self,
+    ///         config: PluginConfiguration,
+    ///         cancellation_token: tedge_api::CancellationToken,
+    ///         core_comms: &PD,
+    ///     ) -> Result<BuiltPlugin, tedge_api::error::PluginError>
+    ///     where
+    ///         PD: 'async_trait,
+    ///     {
+    ///         use tedge_api::plugin::PluginExt;
+    ///         let p = MyPlugin {};
+    ///         Ok(p.into_untyped::<(MyMessage,)>())
+    ///     }
+    ///     // other trait functions...
+    /// #   fn kind_name() -> &'static str {
+    /// #       unimplemented!()
+    /// #   }
+    /// #   fn kind_message_types() -> tedge_api::plugin::HandleTypes
+    /// #   where
+    /// #       Self: Sized,
+    /// #   {
+    /// #       tedge_api::plugin::HandleTypes::declare_handlers_for::<(MyMessage,), MyPlugin>()
+    /// #   }
+    /// #   async fn verify_configuration(
+    /// #       &self,
+    /// #       _config: &PluginConfiguration,
+    /// #   ) -> Result<(), tedge_api::error::PluginError> {
+    /// #       unimplemented!()
+    /// #   }
+    /// }
+    /// ```
+    async fn instantiate(
+        &self,
+        config: PluginConfiguration,
+        cancellation_token: crate::CancellationToken,
+        core_comms: &PD,
+    ) -> Result<BuiltPlugin, PluginError>
+    where
+        PD: 'async_trait;
+}
```

Comment:
From a plugin developer perspective, this `cancellation_token` is my main concern. As a plugin developer, I would prefer not to be responsible for monitoring cancellation (using a `select!`) and have only to deal with the shutdown itself once the cancellation has been acknowledged.

Reply:
I think we discussed this in-voice, but written like this it seems that it is a slightly different topic.

Could you please explain what the difference between the CancellationToken being cancelled and the application shutting down is?